manual/manual/tutorials/coreexamples.etex

```diff
@@ -289,6 +289,91 @@ let rec insert x btree =
+type last   = { x:int };;
+type first_variant = A | B | C
+type last_variant  = A;;
+\end{caml_example*}
+
+The answer is that when confronted with multiple options, OCaml tries to
+use locally available information to disambiguate between the various fields
+and constructors. First, if the type of the record or variant is known,
+OCaml can pick unambiguously the corresponding field or constructor.
+For instance:
+
+\begin{caml_example}{toplevel}
+let look_at_x_then_y_then_z (r:first) =
+  let x = r.x in
+  x + r.z;;
+let rotate_abc_to_bac (x:first_variant) = match x with
+  | A -> (B:first_variant)
+  | B -> A
+  | C -> C;;
+type wrapped = First of first
+let f (First r) = r, r.x;;
+\end{caml_example}
+
+In the first example, "(r:first)" is an explicit annotation telling OCaml
+that the type of "r" is "first". With this annotation, Ocaml knows that
+"r.x" refers to the "x" field of the first type. Similarly, the type
+annotation in the second example makes it clear to OCaml that the
+constructors "A", "B" and "C" come from the first variant type.
+Contrarily, in the last example, OCaml has inferred by itself that the
+type of "r" can only be first and there are no needs for explicit type
+annotations.
+
+Those explicit type annotations can in fact be used anywhere.
+Most of the time they are unnecessary, but they are useful to guide
+disambiguation, to debug unexpected type errors, or combined with some
+of the more advanced features of OCaml described in later chapters.
+
+Secondly, for records, OCaml can also deduce the right record type by
+looking at the whole set of fields used in a expression or pattern:
+\begin{caml_example}{toplevel}
+let project_and_rotate {x;y; _ } = { x= - y; y = x ; z = 0} ;;
+\end{caml_example}
+Since the fields "x" and "y" can only appear simultaneously in the first
+type, OCaml infers that the type of "project_and_rotate" is
+"first -> first".
+
+In last resort, if there is not enough information to disambiguate between
+different fields or constructors, Ocaml picks the last defined type
+amongst all locally valid choices:
+
+\begin{caml_example}{toplevel}
+let look_at_xz {x;z} = x;;
+\end{caml_example}
+
+Here, OCaml has inferred that the possible choices for the type of
+"{x;z}" are "first" and "middle", since the type "last" has no field "z".
+Ocaml then picks the type "middle" as the last defined type between the
+two possibilities.
+
+Beware that this last resort disambiguation is local: once Ocaml has
+chosen a disambiguation, it sticks to this choice, even if it leads to
+an ulterior type error:
+
+\begin{caml_example}{toplevel}[error]
+let look_at_x_then_y r =
+  let x = r.x in (* Ocaml deduces [r: last] *)
+  x + r.y;;
+let is_a_or_b x = match x with
+  | A -> true (* OCaml infers [x: last_variant] *)
+  | B -> true;;
+\end{caml_example}
+
+Consequently, adding explicit type annotations to guide disambiguation is
+more robust than relying on the last defined type disambiguation.
```

Comment:
Another thing to point out is that disambiguation using last declaration is not robust to introducing new declarations or opening modules, which means that the disambiguation choice may change as the code evolve, unlike disambiguation based on type information.

Reply:
I added a short paragraph on this matter.